docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,476 @@
+
+- Limited support for run-time claims
+  The notion of run-time claims is roughly equivalent to the custom claims
+  in the OE attestation plugin framework. The `oe_get_attestation_certificate_with_evidence_v2`
+  API does not allow developers to provide customized custom claims as input.
+  Instead, the API internally takes the input public key buffer as the custom claim.
+  The `oe_verify_attestation_certificate_with_evidence_v2` then extracts the custom
+  claim and verifies its value against the public key.
+
+Proposal
+--------
+
+This document proposes new APIs,
+`oe_get_attestation_certificate_with_evidence_v3` and
+`oe_verify_attestation_certificate_with_evidence_v3`, that
+address the abovementioned limitations. Note that the certificate
+generated by `oe_get_attestation_certificate_with_evidence_v3`
+does not provide full backward compatibility to older version
+of verification APIs (e.g., `oe_verify_attestation_certificate_with_evidence_v2`
+and `oe_verify_attestation_certificate`).
+
+- oe_get_attestation_certificate_with_evidence_v3
+  ```c
+  /**
+   * oe_get_attestation_certificate_with_evidence_v3
+   *
+   * Similar to oe_get_attestation_certificate_with_evidence_2, this function
+   * generates a self-signed x.509 certificate with embedded evidence generated by
+   * an attester plugin for the enclave, but it also allows a user to pass in
+   * runtime and inittime claims.
+   *
+   * @experimental
+   *
+   * @param[in] format_id The format id of the evidence to be generated.
+   *
+   * @param[in] subject_name a string containing an X.509 distinguished
+   * name (DN) for customizing the generated certificate. This name is also used
+   * as the issuer name because this is a self-signed certificate
+   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details
+   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
+   *
+   * @param[in] private_key a private key used to sign this certificate
+   * @param[in] private_key_size The size of the private_key buffer
+   * @param[in] public_key a public key used as the certificate's subject key
+   * @param[in] public_key_size The size of the public_key buffer.
+   * @param[in] inittime_claims an optional flat buffer representing the inittime
+   * claims.
+   * @param[in] inittime_claims_size The size of the inittime_claims in bytes.
+   * @param[in] runtime_claims an optional flat buffer representing the runtime
+   * claims.
+   * @param[in] runtime_claims_size The size of the runtime_claims in bytes.
+   * @param[in] optional_parameters The optional format-specific input parameters.
+   * @param[in] optional_parameters_size The size of optional_parameters in bytes.
+   *
+   * @param[out] output_certificate a pointer to buffer pointer
+   * @param[out] output_certificate_size size of the buffer above
+   *
+   * @return OE_OK on success
+   */
+  oe_result_t oe_get_attestation_certificate_with_evidence_v3(
+      const oe_uuid_t* format_id,
```

Comment:
missing `flags` param that exists for oe_get_evidence().
I will also observe that there is no `endorsements_buffer` output like get_evidence has, why is that?

Reply:
The evidence embedded in the certificate uses the  `OE_EVIDENCE_FLAGS_EMBED_FORMAT_ID` flag by default, which is the contract between the get and verify certificate APIs.

I'm not sure why endorsements is not supported (currently the get API passes endorsements  as `NULL`  to `oe_get_evidence`). I'm also unclear about the use case of supporting endorsements in these APIs.
@yentsanglee can you comment on this?

Reply:
endorsements_buffer is a special contract between evidence generating and verifying. Verifier can use the same endorsement retrieved at the time of evidence generation to verify the evidence. This is useful for verifying possibly stale evidence. I don't see a use case for certificate.

Reply:
Use case is the same as for evidence.